src/v/security/jwt.h

```diff
@@ -0,0 +1,117 @@
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "json/document.h"
+#include "oncore.h"
+#include "outcome.h"
+#include "utils/utf8.h"
+
+#include <optional>
+#include <string_view>
+#include <system_error>
+
+namespace security::oidc {
+
+enum errc {
+    success = 0,
+    metadata_invalid,
+};
+
+struct errc_category final : public std::error_category {
+    const char* name() const noexcept final { return "security::oidc::errc"; }
+
+    std::string message(int c) const final {
+        switch (static_cast<errc>(c)) {
+        case errc::success:
+            return "Success";
+        case errc::metadata_invalid:
+            return "Invalid metadata";
+        }
+    }
+};
+
+inline const std::error_category& error_category() noexcept {
+    static errc_category e;
+    return e;
+}
+
+inline std::error_code make_error_code(errc e) noexcept {
+    return {static_cast<int>(e), error_category()};
+}
+
+} // namespace security::oidc
+
+namespace std {
+template<>
+struct is_error_code_enum<security::oidc::errc> : true_type {};
+
+} // namespace std
+
+namespace security::oidc {
+
+namespace detail {
+
+template<typename CharT = std::string_view::value_type>
+std::basic_string_view<CharT> as_string_view(json::Value const& v) {
+    expression_in_debug_mode(
+      vassert(
+        v.IsString(), "only Strings can be converted to std::string_view"););
+    return {reinterpret_cast<CharT const*>(v.GetString()), v.GetStringLength()};
+}
+
+template<typename CharT = std::string_view::value_type>
+std::optional<std::basic_string_view<CharT>>
+string_view(json::Value const& doc, std::string_view field) {
+    auto it = doc.FindMember(field.data());
+    if (it == doc.MemberEnd() || !it->value.IsString()) {
+        return std::nullopt;
+    }
+    return as_string_view<CharT>(it->value);
+}
+
+} // namespace detail
+
+// Authorization Server Metadata as defined by
+// https://www.rfc-editor.org/rfc/rfc8414.html
+class metadata {
+public:
+    static result<metadata> make(std::string_view sv) {
+        json::Document doc;
+        doc.Parse(sv.data(), sv.length());
+        return make(std::move(doc));
+    }
+    static result<metadata> make(json::Document doc) {
+        if (doc.HasParseError() || !doc.IsObject()) {
+            return errc::metadata_invalid;
+        }
+        for (auto const& field : {"issuer", "jwks_uri"}) {
+            auto f = detail::string_view(doc, field);
+            if (!f || f->empty()) {
+                return errc::metadata_invalid;
+            }
+        }
+        return metadata(std::move(doc));
+    }
+
+    auto issuer() const {
+        return detail::string_view(_metadata, "issuer").value();
+    }
+    auto jwks_uri() const {
+        return detail::string_view(_metadata, "jwks_uri").value();
+    }
```

Comment:
comment: Reading the 8414 rfc, looks like `jwks_uri` is _technically_ optional:

> jwks_uri
      OPTIONAL.  URL of the authorization server's JWK Set [[JWK](https://www.rfc-editor.org/rfc/rfc8414#ref-JWK)]
      document.  The referenced document contains the signing key(s) the
      client uses to validate signatures from the authorization server.
      This URL MUST use the "https" scheme.  The JWK Set MAY also
      contain the server's encryption key or keys, which are used by
      clients to encrypt requests to the server.  When both signing and
      encryption keys are made available, a "use" (public key use)
      parameter value is REQUIRED for all keys in the referenced JWK Set
      to indicate each key's intended usage.